metriq-api/service/platformService.js

```diff
@@ -0,0 +1,207 @@
+      '  RIGHT JOIN "submissionTaskRefs" on submissions.id = "submissionTaskRefs"."submissionId" ' +
+      '  RIGHT JOIN results on results."submissionTaskRefId" = "submissionTaskRefs".id AND (results."deletedAt" IS NULL) ' +
+      '  RIGHT JOIN "resultPlatformRefs" on "resultPlatformRefs"."resultId" = results.id AND ("resultPlatformRefs"."deletedAt" IS NULL) ' +
+      '  RIGHT JOIN platforms on platforms.id = "resultPlatformRefs"."platformId" ' +
+      '  WHERE platforms.id = ' + platformId
+    ))[0][0].count
+  }
+
+  async getPropertiesByPk (platformId) {
+    return (await sequelize.query(
+      'SELECT "platformDataTypeValues".id as id, "platformDataTypes".name AS name, "platformDataTypes"."dataTypeId" as "dataTypeId", "dataTypes".name AS type, "dataTypes"."friendlyName" AS "typeFriendlyName", "platformDataTypeValues".value AS value FROM platforms ' +
+      '  LEFT JOIN "platformDataTypeValues" on platforms.id = "platformDataTypeValues"."platformId" ' +
+      '  LEFT JOIN "platformDataTypes" on "platformDataTypes".id = "platformDataTypeValues"."platformDataTypeId" ' +
+      '  LEFT JOIN "dataTypes" on "platformDataTypes"."dataTypeId" = "dataTypes".id ' +
+      '  WHERE platforms.id = ' + platformId
+    ))[0]
+  }
+
+  async getTopLevelNamesAndCounts () {
+    const result = (await this.getAllNames()).body
+    for (let i = 0; i < result.length; i++) {
+      result[i].dataValues.submissionCount = parseInt(await this.getSubmissionCount(result[i].id))
+      result[i].dataValues.upvoteTotal = parseInt(await this.getLikeCount(result[i].id))
+      result[i].dataValues.resultCount = parseInt(await this.getResultCount(result[i].id))
+    }
+    const filtered = []
+    for (let i = 0; i < result.length; i++) {
+      if (result[i].submissionCount > 0) {
+        filtered.push(result[i])
+      }
+    }
+    return { success: true, body: result }
+  }
+
+  async getByName (name) {
+    return await this.SequelizeServiceInstance.findOne({ name: name })
+  }
+
+  async getAllNames () {
+    const result = await this.SequelizeServiceInstance.projectAll(['id', 'name'])
+    return { success: true, body: result }
+  }
+
+  async submit (userId, reqBody) {
+    const nameMatch = await this.getByName(reqBody.name)
+    if (nameMatch) {
+      return { success: false, error: 'Platform name already in use.' }
+    }
+
+    let platform = await this.SequelizeServiceInstance.new()
+    platform.userId = userId
+    platform.name = reqBody.name
+    platform.fullName = reqBody.fullName
+    platform.description = reqBody.description
+    platform.platformId = reqBody.parentPlatform ? reqBody.parentPlatform : null
+
+    if (reqBody.parentPlatform) {
+      const parentPlatform = await this.getByPk(platform.platformId)
+      if (!parentPlatform) {
+        return { success: false, error: 'Parent platform ID does not exist.' }
+      }
+    }
+
+    // Get an ObjectId for the new object, first.
```

Comment:
Is this comment relevant to the code block below? Perhaps it is but I'm just not able to see it.